src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -2902,12 +2886,29 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           }
         }
 
-        // if we had missing param types, see if we can undo eta-expansion and recover type info
-        val expectedFunTypeBeforeEtaExpansion =
-          if (paramsMissingType.isEmpty) NoType
-          else typeUnEtaExpanded
+        // If we're typing `(a1: T1, ..., aN: TN) => m(a1,..., aN)`, where some Ti are not fully defined,
+        // type `m` directly (undoing eta-expansion of method m) to determine the argument types.
+        val ptUnrollingEtaExpansion =
+          if (paramsMissingType.nonEmpty && pt != ErrorType) fun.body match {
+            case Apply(meth, args) if (vparams corresponds args) { case (p, Ident(name)) => p.name == name case _ => false } =>
```

Comment:
Took me a while to convince myself that this (existing) code was correct, usually matching on names like this is sketchy.

But there is no way that an `Ident` argument to the method can bind to anything other than the eponymous function parameter. 

```scala
{ val a = ""; (a, b) => bar(a /*Ident(a) can't be the outer a*/, b) }
```